### PR TITLE
refactor: async iterators

### PR DESCRIPTION
**Pull request by alanshaw**
@jacobheun this is https://github.com/alanshaw/it-mplex + the [example from this repo](https://github.com/libp2p/js-libp2p-mplex/tree/master/examples) updated to use the new API.

Note this includes an adapter class so it can be used immediately in JS IPFS when merged and released. I've run the JS IPFS tests and interop tests against it and they pass!

TODO:

* [x] Fix up the one failing test from https://github.com/libp2p/interface-stream-muxer/pull/55 (either here or there 😂)

Depends on:

* [x] https://github.com/libp2p/interface-stream-muxer/pull/55

BREAKING CHANGE: This module now has a completely new API based on async iterators. See the README for documentation.

### CODE REVIEW

**Comment by jacobheun** [comment]
@alanshaw the restrictSize fix took care of the issue with benchmarks. When I was running that the restrict size error ends up getting swallowed and never logged, only abort errors show in the logs. Can we have the source errors propagate up to the logs, they seem to be getting swallowed?

**Comment by alanshaw** [comment]
I'll look into it - this still needs work clearly. The slow pingpong benchmark is also worrying.

**Comment by jacobheun** [comment]
I used npm alias support to install the interface-stream-muxer@0.6 code along side the async test suite to test the adapter, as I was curious about the close tests. It's failing the close tests there as well (both close tests fail for the adapter), so I think there is something wrong with error/close propagation.

**Comment by alanshaw** [comment]
@jacobheun I'm working on a fork of the `libp2p-tcp` async/await lib that's used in that test. I built https://www.npmjs.com/package/stream-to-it to ensure all cases were covered when a writable stream closes or errors. When `libp2p-tcp` is ready the test should pass and we should be gtg. My plan is then to circle back to the `interface-connection` PR.

File: src/coder/decode.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+    this._buffer.append(chunk)
+    const msgs = []
+
+    while (true) {
+      if (!this._headerInfo) {
+        try {
+          this._headerInfo = this._decodeHeader(this._bufferProxy)
+        } catch (_) {
+          break // not enough data yet...probably
```
Comment (daviddias) [inline src/coder/decode.js]: Can you elaborate a bit more? The "probably" is not really comforting :)

**Comment by jacobheun** [comment]
[js-libp2p-mplex migration is complete](https://trello.com/c/ZgwcWzur/1-js-libp2p-mplex-migration-is-complete)

File: src/coder/decode.js (hunk 1 above)
Comment (jacobheun) [reply src/coder/decode.js]: Updated the comment for this. We should only hit the break in 2 instances. 
1. We haven't gotten enough data to parse a header (most common).
2. The data we have can't be parsed, likely due to an invalid varint from bad data.

**Comment by jacobheun** [comment]
Fixed, the BufferList chunks need to be converted before going to node streams .write

**Pull request merged**